src/Lib.P2P.Tests/SwarmTest.cs

```diff
@@ -0,0 +1,1277 @@
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using MultiFormats;
+
+namespace Lib.P2P.Tests
+{
+    [TestClass]
+    public sealed class SwarmTest
+    {
+        private readonly MultiAddress _mars =
+            "/ip4/10.1.10.10/tcp/29087/ipfs/QmSoLMeWqB7YGVLJN3pNLQpmmEk35v6wYtsMGLzSr5QBU3";
+
+        private readonly MultiAddress _venus =
+            "/ip4/104.236.76.40/tcp/4001/ipfs/QmSoLV4Bbm51jM9C4gDYZQ9Cy3U6aXMJDAbzgu2fzaDs64";
+
+        private readonly MultiAddress _earth =
+            "/ip4/178.62.158.247/tcp/4001/ipfs/QmSoLer265NRgSp2LA3dPaeykiS1J6DifTC88f5uVQKNAd";
+
+        private readonly Peer _self = new Peer
+        {
+            AgentVersion = "self",
+            Id = "QmXK9VBxaXFuuT29AaPUTgW3jBWZ9JgLVZYdMYTHC6LLAH",
+            PublicKey =
+                "CAASXjBcMA0GCSqGSIb3DQEBAQUAA0sAMEgCQQCC5r4nQBtnd9qgjnG8fBN5+gnqIeWEIcUFUdCG4su/vrbQ1py8XGKNUBuDjkyTv25Gd3hlrtNJV3eOKZVSL8ePAgMBAAE="
+        };
+
+        private readonly Peer _other = new Peer
+        {
+            AgentVersion = "other",
+            Id = "QmdpwjdB94eNm2Lcvp9JqoCxswo3AKQqjLuNZyLixmCM1h",
+            PublicKey =
+                "CAASXjBcMA0GCSqGSIb3DQEBAQUAA0sAMEgCQQDlTSgVLprWaXfmxDr92DJE1FP0wOexhulPqXSTsNh5ot6j+UiuMgwb0shSPKzLx9AuTolCGhnwpTBYHVhFoBErAgMBAAE="
+        };
+
+        [TestMethod]
+        public async Task Start_Stop()
+        {
+            var swarm = new SwarmService
+            {
+                LocalPeer = _self
+            };
+            await swarm.StartAsync();
+            await swarm.StopAsync();
+        }
+
+        [TestMethod]
+        public void Start_NoLocalPeer()
+        {
+            var swarm = new SwarmService();
+            ExceptionAssert.Throws<NotSupportedException>(() =>
+            {
+                swarm.StartAsync().Wait();
+            });
+        }
+
+        [TestMethod]
+        public void NewPeerAddress()
+        {
+            var swarm = new SwarmService {LocalPeer = _self};
+            swarm.RegisterPeerAddress(_mars);
+            Assert.IsTrue(swarm.KnownPeerAddresses.Contains(_mars));
+        }
+
+        [TestMethod]
+        public void NewPeerAddress_Self()
+        {
+            var swarm = new SwarmService {LocalPeer = _self};
+            var selfAddress = "/ip4/178.62.158.247/tcp/4001/ipfs/" + _self.Id;
+            ExceptionAssert.Throws<Exception>(() =>
+            {
+                var _ = swarm.RegisterPeerAddress(selfAddress);
+            });
+
+            selfAddress = "/ip4/178.62.158.247/tcp/4001/p2p/" + _self.Id;
+            ExceptionAssert.Throws<Exception>(() =>
+            {
+                var _ = swarm.RegisterPeerAddress(selfAddress);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove the unused local variable '_'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)